### PR TITLE
Redo the docs for Vec::set_len

### PR DESCRIPTION
**Pull request by scottmcm**
Inspired by the [recent conversation on IRLO](https://internals.rust-lang.org/t/make-vec-set-len-enforce-the-len-cap-invariant/8927/23?u=scottmcm).

This is just my first stab at this; suggestions welcome.

### CODE REVIEW

File: src/liballoc/vec.rs [hunk 1]
```diff
@@ -748,28 +748,64 @@ impl<T> Vec<T> {
         self
     }
 
-    /// Sets the length of a vector.
+    /// Forces the length of a vector to a particular value.
```
Comment (Centril) [inline src/liballoc/vec.rs]:
```suggestion
    /// Forces the length of a vector to be the given value in `new_len`.
```
particular doesn't sound very particular. =P

File: src/liballoc/vec.rs [hunk 2]
```diff
@@ -748,28 +748,64 @@ impl<T> Vec<T> {
...
     ///
-    /// This will explicitly set the size of the vector, without actually
-    /// modifying its buffers, so it is up to the caller to ensure that the
-    /// vector is actually the specified size.
+    /// This is a low-level operation that maintains none of the normal
+    /// invariants of the type.  Normally changing the length of a `Vec`
```
Comment (Centril) [inline src/liballoc/vec.rs]:
```suggestion
    /// invariants of the type.  Normally changing the length of a `Vec<T>`
```

File: src/liballoc/vec.rs [hunk 3]
```diff
@@ -748,28 +748,64 @@ impl<T> Vec<T> {
...
+    /// is done using one of the safe operations instead, such as
```
Comment (Centril) [inline src/liballoc/vec.rs]:
```suggestion
    /// is done using one of the *safe operations* instead, such as
```

File: src/liballoc/vec.rs [hunk 4]
```diff
@@ -748,28 +748,64 @@ impl<T> Vec<T> {
...
+    /// [`resize`]: #method.resize
+    /// [`extend`]: #method.extend-1
+    /// [`clear`]: #method.clear
     ///
-    /// ```
-    /// use std::ptr;
+    /// # Safety
     ///
-    /// let mut vec = vec!['r', 'u', 's', 't'];
+    /// - `new_len` must be less than or equal to `capacity()`.
```
Comment (Centril) [inline src/liballoc/vec.rs]: Hyperlink `capacity()`

Comment (Centril) [inline src/liballoc/vec.rs]:
```suggestion
    /// - `new_len` must be less than or equal to the vector's `capacity()`.
```
Since we are stating safety invariants being extra clear could be better.

File: src/liballoc/vec.rs [hunk 5]
```diff
@@ -748,28 +748,64 @@ impl<T> Vec<T> {
...
+    /// - All elements between past the previous end up to the `new_len`
```
Comment (Centril) [inline src/liballoc/vec.rs]: It's hard to parse exactly what  "between past the previous end up to" means, particularly true if you are not a native English speaker. Can we give a backup wording that relies on a more formal and universal language (i.e. stating it in math/Rust-like notation)?

File: src/liballoc/vec.rs [hunk 6]
```diff
@@ -748,28 +748,64 @@ impl<T> Vec<T> {
...
+    ///   must be initialized.
     ///
-    /// unsafe {
-    ///     ptr::drop_in_place(&mut vec[3]);
-    ///     vec.set_len(3);
+    /// # Examples
+    ///
+    /// This method can be useful for situations in which the `Vec` is
```
Comment (Centril) [inline src/liballoc/vec.rs]:
```suggestion
    /// This method can be useful for situations in which the `Vec<T>` is
```

File: src/liballoc/vec.rs [hunk 7]
```diff
@@ -748,28 +748,64 @@ impl<T> Vec<T> {
...
+    /// # extern "C" {
+    /// #     fn deflateGetDictionary(
+    /// #         strm: *mut std::ffi::c_void,
+    /// #         dictionary: *mut u8,
+    /// #         dictLength: *mut usize,
+    /// #     ) -> i32;
+    /// # }
+    /// # impl StreamWrapper {
+    /// pub fn get_dictionary(&self) -> Option<Vec<u8>> {
+    ///     // Per the docs, "32768 bytes is always enough".
```
Comment (Centril) [inline src/liballoc/vec.rs]: What docs; that of the C FFI or that of `Vec<T>`? Reword to make it clearer which one it is?

File: src/liballoc/vec.rs [hunk 8]
```diff
@@ -748,28 +748,64 @@ impl<T> Vec<T> {
...
+    ///     let mut dict = Vec::with_capacity(32_768);
+    ///     let mut dict_length = 0;
+    ///     unsafe {
```
Comment (Centril) [inline src/liballoc/vec.rs]: Imo, we should attach a comment to `unsafe { ... }` explaining why it is safe to encourage providing such explanations in the ecosystem.

File: src/liballoc/vec.rs [hunk 9]
```diff
@@ -748,28 +748,64 @@ impl<T> Vec<T> {
...
+    ///     }
     /// }
-    /// assert_eq!(vec, ['r', 'u', 's']);
+    /// # }
     /// ```
     ///
-    /// In this example, there is a memory leak since the memory locations
-    /// owned by the inner vectors were not freed prior to the `set_len` call:
+    /// While the following example is sound, there is a memory leak since
+    /// the inner vectors were not freed prior to the `set_len` call:
```
Comment (Centril) [inline src/liballoc/vec.rs]: It would be good to add a comment to the following `unsafe { ... }` block as well.

File: src/liballoc/vec.rs [hunk 10]
```diff
@@ -780,21 +816,11 @@ impl<T> Vec<T> {
...
-    /// values of unallocated memory:
-    ///
-    /// ```
-    /// let mut vec: Vec<char> = Vec::new();
-    ///
-    /// unsafe {
-    ///     vec.set_len(4);
-    /// }
-    /// ```
+    /// (Instead, one would normally use [`clear`] in this situation.)
```
Comment (Centril) [inline src/liballoc/vec.rs]: No need for a parenthetical imo; it's important info.

**Review by Centril — COMMENTED** [review]
Some improvement suggestions. :)

File: src/liballoc/vec.rs (hunk 1 above)
Comment (nagisa) [reply src/liballoc/vec.rs]: > Forces the length of the vector to `new_len`

seems less mouthful to me.

File: src/liballoc/vec.rs (hunk 5 above)
Comment (nagisa) [reply src/liballoc/vec.rs]: > All elements between `Vec::len()` and `new_len` must be initialized by the user after `set_len` returns

seems like a better wording, as it describes *when* the initialization should occur. There’s a slight ambiguity in the wording (both currently PRd and my proposed) that one could interpret

```rust
fn weird_fn<T>(v: &mut Vec<T>) { 
    let old_len = v.len();
    v.set_len(0);
    // ...
    v.set_len(old_len);
}
```

as invalid… would be great to find a wording that does not have such ambiguity.

File: src/liballoc/vec.rs (hunk 10 above)
Comment (nagisa) [reply src/liballoc/vec.rs]: Perhaps it should be made more clear that `clear` differs from `set_len` exactly in how they behave wrt dropping the contents?

**Comment by Dylan-DPC** [comment]
ping from triage @scottmcm can you update this with the changes suggested above? 
r?@centril

File: src/liballoc/vec.rs (hunk 1 above)
Comment (Centril) [reply src/liballoc/vec.rs]: Updated the suggestion to @nagisa's wording which is more direct / better.

File: src/liballoc/vec.rs (hunk 5 above)
Comment (scottmcm) [reply src/liballoc/vec.rs]: I tried range notation; let me know if that's better.

File: src/liballoc/vec.rs (hunk 8 above)
Comment (scottmcm) [reply src/liballoc/vec.rs]: I'm not super convinced of that, actually, because of tootsie pop.  A bunch of the justifications are actually invariants of the type+module, and repeating "safe because the type invariant" all over is questionably useful, as is "safe because we trust the FFI method".

I _did_ comment the two unsafe _calls_.  If you'd like more on the block, propose some text that's more helpful than distracting.

File: src/liballoc/vec.rs (hunk 5 above)
Comment (Centril) [reply src/liballoc/vec.rs]: Yeah, much better!

File: src/liballoc/vec.rs (hunk 9 above)
Comment (Centril) [reply src/liballoc/vec.rs]: Something along the lines of: "Safe because 1) 0 is trivially <= any capacity and 2) N..0 is an empty range so no elements need to be initialized."

**Comment by scottmcm** [comment]
@Centril I'm still on vacation without access to my computer.  Can you make proposed changes here so I can apply them, like I just did your other one?

**Comment by Centril** [comment]
@scottmcm sure, I'll get to it in a bit.

**Comment by Centril** [comment]
@scottmcm made a PR instead: https://github.com/scottmcm/rust/pull/1

@bors delegate+

**Comment by pietroalbini** [comment]
@bors retry
AppVeyor... what's wrong with you today?
